### DISCUSSION TITLE
Desire behaviour for dot graph import

### DISCUSSION THREAD

**Discussion by KelvinChung2000** [networkx/networkx#7709] [Q&A] [answered]
This is a reference to #7698 

When importing the following graph:

```dot
strict graph  {
graph [test=test];
graph [test1=test2];
node [a=0]
node [b=1]
edge [c=2]
edge [d=3]
}
```

The result graph attribute will be as follows:
```python
import networkx as nx

G = nx.drawing.nx_pydot.read_dot("test.dot")
print(G.graph)

# {'name': 'G', 'node': {'a': '0'}, 'edge': {'c': '2'}}

```

However, I think the following makes more sense:

```python
import networkx as nx

G = nx.drawing.nx_pydot.read_dot("test.dot")
print(G.graph)

# {'name': 'G', 'test': 'test', 'test1': 'test2', 'node': {'a': '0', 'b': '1'}, 'edge': {'c': '2', 'd': '3'}}
```

or even the `node` and `edge` should not be in the `graph` attribute, and the value is automatically copied as the node and edge data. 

The current behaviour is the following:

```python
import networkx as nx

G = nx.Graph(graph={"test": "test"})
nx.drawing.nx_pydot.write_dot(G, "test.dot")
``` 

To output:
```dot
strict graph {
test=test;
}
```

and currently only accept the `test=test;` syntax when importing the graph. 

I am adding support to the `graph[key=value];` syntax, but should we keep the `G.graph ["graph"]` to get the graph data or use `G.graph` to see all the graph data? Similarly, should we copy the `G.graph ["node"]` and `G.graph ["edge"]` data to all the nodes and edges?

**Comment by dschult** [comment]
I think a good interface to follow would be the output of `nx.nx_agraph.write_dot(G, path)` when that makes sense --- I'm not saying it is foolproof; there may be improvements that could be made. But it gives a general guideline to follow.

The `agraph` interface through pygraphviz uses GraphViz libraries to write the dot file. So it should follow the GraphViz standards which is what we care about most. These dot files should work as expected with GraphViz.

I'm not sure if that description answers your questions. It seems like all the data attributes should be loaded onto the nodes, edges and the `G.graph` attribute dict.  For example, I would not expect to find a "graph" attribute inside the `G.graph` dict. Nor should `G.graph` include any node or edge information. `G.graph` is intended for meta data about the graph rather than info about the nodes/edges/structure of the graph. But I don't know how that interacts with the pydot interface and it's interaction with GraphViz.

**Reply by KelvinChung2000** [reply]
> I would not expect to find a "graph" attribute inside the G.graph dict. Nor should G.graph include any node or edge information. G.graph is intended for meta data about the graph rather than info about the nodes/edges/structure of the graph. 

I agree, and it makes sense to me.  

I mainly use `pydot` because this is the first answer when you Google "how to export networkx graph to dot", and I don't know `agraph` does the same thing. 

I used the same example with `agraph`, and it gives the following:

```python
import networkx as nx

G = nx.Graph(test="test")
nx.drawing.nx_agraph.write_dot(G, "test.dot")
G = nx.drawing.nx_agraph.read_dot("test.dot")

print(G.graph)
# {'test': 'test', 'graph': {'test': 'test'}, 'node': {}, 'edge': {}}
```

test.dot:
```dot
strict graph "" {
	graph [test=test];
}
```

which is also not very intuitive. 

Making things match will probably take a while since I will also need to spend time learning the `pydot` interface.  I will need to rewrite a lot of the test as well, which is quite a lot of work. 

If the `agraph` interface should be the "standard", will deprecating the `pydot` `read_dot` and `write_dot` make sense? The `agraph` interface will be the only way to read and write a DOT file. The only argument I would make against deprecating the `pydot` interface would be that `pydot` does not require any system package installation.

**Comment by dschult** [comment, marked as answer]
It looks like pydot and pygraphviz both give very close to the same dot file. So I'm not sure that anything is needed at the moment. Intuitive or not, that is how GraphViz expects dot-files to encode the information.
